enhancements/machine-api/cluster-autoscaler-operator.md

```diff
@@ -0,0 +1,353 @@
+
+To opt particular `MachineSets` into autoscaling, administrators will create `MachineAutoscaler` custom resources.
+These resources will reference and provide scaling limits for a single `MachineSet`.
+The Cluster Autoscaler Operator will use these `MachineAutoscalers` to configure the Cluster Autoscaler to autoscale
+the referenced `MachineSet`.
+
+### User Stories
+
+#### Story 1
+
+As a user of OpenShift, it should be easy to enable autoscaling of compute resources so that my cluster will respond to
+increased usage without manual intervention.
+
+#### Story 2
+
+As a user of OpenShift, I should be able to specify which groups of Machines should be autoscaled and which should not.
+
+#### Story 3
+
+As a user of OpenShift, I should be able to limit the size of my cluster when it is autoscaling so that I do not overload
+the Control Plane and so that I can cap the costs of my cluster.
+
+#### Story 4
+
+As a user of OpenShift, I should be able to specify a maximum and minimum size for any group of Machines being scaled
+so that I can ensure I have a minimum capacity available and that I do not over commit to a single pool of Machines.
+
+### Implementation Details
+
+The controller will consist of two control loops, one for a new `ClusterAutoscaler` CRD, and the second for a new
+`MachineAutoscaler` CRD.
+
+This controller will be deployed and managed by the Cluster Version Operator and will become a second level operator.
+
+#### ClusterAutoscaler controller
+
+The `ClusterAutoscaler` controller is responsible for managing Cluster Autoscaler Deployments in the `openshift-machine-api` namespace.
+It will also be responsible for ensuring that the relevant resources for providing monitoring information will also be deployed.
+
+If a `ClusterAutoscaler` resource exists, the controller will perform the following logic:
+- Fetch the `ClusterAutoscaler` custom resource
+- Validate the options specified on the `ClusterAutoscaler` specification
+  - Should the validation fail, log an error and requeue the custom resource
+- Check for a Deployment for the `ClusterAutoscaler` custom resource
+  - A Deployment named `cluster-autoscaler-<name>` would be expected, where `<name>` is the name of the `ClusterAutoscaler` resource
+- Ensure monitoring resources are set up for the Cluster Autoscaler deployment
+  - This includes a `Service`, `ServiceMonitor` and `PrometheusRule`
+- If the Deployment does not exist, create it
+- If the Deployment exists, ensure the Pod specification matches the desired configuration from the `ClusterAutoscaler` resource
+
+#### ClusterAutoscaler Admission Webhook
+
+The `ClusterAutoscaler` admission webhook will be responsible for validating the configuration of `ClusterAutoscaler`
+resources as they are created or updated.
+
+It will perform the following validation:
+- Check the `ClusterAutoscaler` name
+  - Only one `clusterAutoscaler` will be allowed per cluster to prevent multiple Cluster Autoscaler deployments from
```

Comment:
should this be `ClusterAutoscaler`, just to keep consistent with the previous usage?